src/Body.ts

```diff
@@ -1,44 +1,590 @@
+				const atStart = (sliceStart === 0);
+				const atEnd = (sliceStart + sliceCount) === max;
+				if ((startDelta || countDelta) && (atStart || atEnd || startDelta > rowDrift || countDelta > rowDrift)) {
+					onSliceRequest && onSliceRequest({ start: sliceStart, count: sliceCount });
+				}
+			}
+		}
+	}
+
+	render(): DNode {
+		const {
+			_itemElementMap: itemElementMap,
+			properties: {
+				items,
+				size: {
+					dataLength
+				},
+				slice: {
+					start = 0
+				} = {}
+			}
+		} = this;
+		const max = (dataLength > 0 ? dataLength - 1 : 0);
+
+		const children: DNode[] = [];
+
+		// Create a top margin if the data has any offset at all
+		let marginTop = this._marginTop;
+		if (start > 0) {
+			if (!marginTop) {
+				marginTop = this._marginTop = {
+					add: true,
+					remove: false,
+					index: -1
+				};
+			}
+			children.push(v('div', {
+				key: 'marginTop',
+				styles: {
+					height: '10000px'
+				}
+			}));
+		}
+		else if (marginTop) {
+			if (marginTop.element) {
+				marginTop.height = marginTop.element.offsetHeight;
+			}
+			marginTop.remove = true;
+		}
+
+		const previousKeys = from(itemElementMap.keys());
+		if (previousKeys.length === 0) {
+			// There were no item rows the last time render was called
+			// so every row is added
+			for (let i = 0, item; (item = items[i]); i++) {
+				children.push(this.createNodeFromItem(item, (start + i)));
+			}
+		}
+		else {
+			// Create a new map so that the items will be ordered correctly
+			const updatedElementMap = new Map<string, RenderedDetails>();
+
+			// Keep a map of current keys (item IDs) and items
+			const itemsByKey: { [key: string]: DNode } = {};
+			const currentKeys: string[] = items.map((item, index) => {
+				const key = item.id;
+				// createNodeFromItem marks this item as having been added
+				// automatically if it didn't exist in the mapping (is new)
+				itemsByKey[key] = this.createNodeFromItem(item, (start + index));
+				return key;
+			});
+
+			// find which keys are new and at what index they will appear
+			let cleared = true;
+			let addedKeys: string[] = [];
+			const keyPatches: { [ index: number]: string[] } = {};
+			let previousKeyIndex = 0;
+			for (const currentKey of currentKeys) {
+				const foundAtIndex = previousKeys.indexOf(currentKey, previousKeyIndex);
+				if (foundAtIndex === -1) {
+					addedKeys.push(currentKey);
+				}
+				else {
+					if (addedKeys.length) {
+						keyPatches[previousKeyIndex] = addedKeys;
+						addedKeys = [];
+					}
+
+					cleared = false;
+					previousKeyIndex = (foundAtIndex + 1);
+				}
+			}
+			if (addedKeys.length) {
+				keyPatches[previousKeys.length] = addedKeys;
+			}
+
+			// If all keys are new, we can start from scratch
+			if (cleared) {
+				itemElementMap.clear();
+				return this.render();
+			}
+
+			// Use previous keys to watch for deleted items
+			// and insert new keys at the indexes detected above
+			for (let i = 0, il = previousKeys.length; i <= il; i++) {
```

Comment:
Can we use a more descriptive name than `il`? Perhaps `i` as well given the length and complexity of this loop.